File: myfunc.py
import base64
import json
import common

def myfunc(queryobj):
    try:
        postdict = queryobj._POST()
        filesdict = queryobj._FILES()
        
        print("POST = " + str(postdict) + "\n")
        print("FILES = " + str(filesdict) + "\n")

        replymsg = ""

        if postdict["request"] == "in1":
            res = "In1-{}".format(postdict["in1"])
            replymsg = json.dumps([postdict["request"],res]).encode('UTF-8')
        #
        elif postdict["request"] == "in2":
            res = "In2-{}".format(postdict["in2"])
            replymsg = json.dumps([postdict["request"],res]).encode('UTF-8')
        #
        elif postdict["request"] == "in3":
            res = "In3-{}".format(postdict["in3"])
            replymsg = json.dumps([postdict["request"],res]).encode('UTF-8')
        #
        elif postdict["request"] == "doc1":

            # reply message should be encoded to be sent back to browser ----------------------------------------------
            # encoding to base64 is used to send ansi hebrew data. it is decoded to become string and put into json.
            # json is encoded to be sent to browser.

            if bool(filesdict):
                file64enc = base64.b64encode(filesdict['doc1'][1])
                file64dec = file64enc.decode()
            
                replymsg = json.dumps([filesdict['doc1'][0],file64dec]).encode('UTF-8')
            #
            else: #if filesdict is empty
                replymsg = json.dumps(["Error","No file provided"]).encode('UTF-8')
            #
        #
        print(replymsg)    
        return replymsg
    #
    
    except Exception as e:
        common.errormsg(title=__name__,message=e)
        replymsg = json.dumps(["Error",__name__+" -" + str(e)]).encode('UTF-8')
        return replymsg
    #
#

File: myfunc.js
myfunc = new Object();
//*********************************************************************************** */
myfunc.msg = function(title,msg_txt){
    document.getElementById("msg_title").innerHTML = title
    document.getElementById("msg_txt").innerHTML = msg_txt
    document.getElementById("msg_dg").showModal()
}
//********************************************************************************************* */
myfunc.response = function(inid,intxt){
    document.getElementById(inid + "_res").innerHTML = intxt
    document.getElementById("response_dg").showModal()
}
//********************************************************************************************* */
myfunc.resp_close = function(){
    document.getElementById("response_dg").close();
    
    resnames = document.getElementsByName("inres")
    
    for (resnm of resnames){
        resnm.innerHTML = ""
    }
}
//********************************************************************************************* */
myfunc.sendrequest = function(fdata){
    return new Promise((resolve) =>{    
        var xhr = new XMLHttpRequest();
        xhr.open('POST',"http://127.0.0.1:"+ui.port,true)
    
        xhr.onreadystatechange = function() {
            if (this.readyState == 4 && this.status == 200) {   
                console.log(this.responseText)
    
                resobj = JSON.parse(this.responseText);
                resolve(resobj)    
     
            }
            else if (this.readyState == 4 && this.status != 200){
                resolve(["Error",this.responseText])

            }
        }
    
        xhr.send(fdata);   
    })
}

//*********************************************************************************** */
myfunc.submit = async function(){ //request can be insert or update

    arr = ["in1","in2","in3","doc1"]

    for (eacharr of arr){
        files = document.getElementById(eacharr).files
        
        if(files != null){
            if (files.length > 0){
                var fdata = new FormData();
                fdata.append("request",eacharr)
                fdata.append(eacharr,document.getElementById(eacharr).files[0]);
            }
        }
        else{
            if (document.getElementById(eacharr).value != ""){
                var fdata = new FormData();
                fdata.append("request",eacharr)
                fdata.append(eacharr,document.getElementById(eacharr).value);;    
                
            }
        }
    
        const resobj = await myfunc.sendrequest(fdata)
        if (resobj[0] == "Error"){
            myfunc.msg(resobj[0], resobj[1])
       }
       else{
           if (resobj[0].slice(0,2) == "in"){
               myfunc.response(resobj[0], resobj[1])
           }
           else {
               myfunc.download(resobj[0], resobj[1])
           }
       } 
    }
    
}

//********************************************************************************************* */
myfunc.download = function(filename, filetext){

    var a = document.createElement("a");

    document.body.appendChild(a);

    a.style = "display: none";

    a.href = 'data:application/octet-stream;base64,' + filetext;

    a.download = filename;

    a.click();

    document.body.removeChild(a);

}
